### PR TITLE
Add some more SPIRE details to TEP0089

### PR DESCRIPTION
**Pull request by lumjjb**
Added a bit more detail about SPIRE, and also added a couple more details based on all the heavy lifting @pxp928 is doing!

Signed-off-by: Brandon Lum <lumjjb@gmail.com>

### CODE REVIEW

File: teps/0089-nonfalsifiable-provenance-support.md [hunk 1]
```diff
@@ -59,26 +63,41 @@ There are a couple issues with this:
...
-SPIRE can be used to request certificates (SVIDs) for a given k8s workload and can use these certificates to create signatures for the data we need to verify, i.e. an entire TaskRun yaml and results. For our use case, both the Tekton entrypointer image (running in Pods) and the Tekton Controller will interact with SPIRE.
+Pods running in a cluster can interact with SPIRE through SPIFFE api via the local SPIRE agent socket.
+
+Getting pod identity
+- A pod can request for an identity by talking to the local agent socket. The SPIRE agent attests the identity of the pod,
+and request for an identity for the pod to the SPIRE server.
+- The workload then receives one or more SPIFFE Verifiable Identity Document, or [SVID](https://spiffe.io/docs/latest/spiffe-about/spiffe-concepts/#spiffe-verifiable-identity-document-svid), an X509 certificate together with the associated private key for the pod's identity. The x509 certificate is signed by the SPIRE server's certificate authority.
+
+Verifying pod identity
+- A pod can request for an trust bundle of a SPIRE server, which will include the public key to validate the pod's certificate.
```
Comment (vdemeester) [inline teps/0089-nonfalsifiable-provenance-support.md:85]: typo: for **a** trust bundle

File: teps/0089-nonfalsifiable-provenance-support.md [hunk 2]
```diff
@@ -59,26 +63,41 @@ There are a couple issues with this:
...
 
 ## SPIRE Concepts
 Before discussing how SPIRE is going to fix these issues, here's a very basic overview of how it works:
 
-Pods running in the cluster can interact with SPIRE.
-A pod will send some `payload` to SPIRE for signing.
-The SPIRE socket will confirm that the request came from an approved workload (in our case, the Tekton controller or Pods created by the controller) and then it will sign the `payload`.
-SPIRE will return:
-1. The `signature` over the `payload`
-1. A SPIFFE Verifiable Identity Document, or [SVID](https://spiffe.io/docs/latest/spiffe-about/spiffe-concepts/#spiffe-verifiable-identity-document-svid), an X509 certificate containing the public key used to verify the `signature` against the `payload`
```
Comment (bobcatfish) [inline teps/0089-nonfalsifiable-provenance-support.md:75]: @lumjjb just double checking, you removed this description of SPIRE signing a payload via the SPIRE socket which I assume means SPIRE does not actually do any signing? i.e. it just make the identity doc (SVID) and the private key available, and the caller is then responsible for what happens after that, e.g. doing the actual signing?

File: teps/0089-nonfalsifiable-provenance-support.md [hunk 3]
```diff
@@ -59,26 +63,41 @@ There are a couple issues with this:
...
+- The SPIRE server acts as local endpoints for each compute node that is part of the SPIRE deployment.
 
-SPIRE can be used to request certificates (SVIDs) for a given k8s workload and can use these certificates to create signatures for the data we need to verify, i.e. an entire TaskRun yaml and results. For our use case, both the Tekton entrypointer image (running in Pods) and the Tekton Controller will interact with SPIRE.
+Pods running in a cluster can interact with SPIRE through SPIFFE api via the local SPIRE agent socket.
+
+Getting pod identity
+- A pod can request for an identity by talking to the local agent socket. The SPIRE agent attests the identity of the pod,
+and request for an identity for the pod to the SPIRE server.
+- The workload then receives one or more SPIFFE Verifiable Identity Document, or [SVID](https://spiffe.io/docs/latest/spiffe-about/spiffe-concepts/#spiffe-verifiable-identity-document-svid), an X509 certificate together with the associated private key for the pod's identity. The x509 certificate is signed by the SPIRE server's certificate authority.
```
Comment (bobcatfish) [inline teps/0089-nonfalsifiable-provenance-support.md:82]: im a bit surprised that SPIRE sends the private key to the caller - is there anything that limits how this private key is used after it has been fetched?

tekton tasks are executed as pods, which in this proposal would have the SPIRE socket mounted into them - it sounds like a user could add steps into their Task (and resulting pod) that could fetch this private key and upload it somewhere else for use later - am i understanding correctly?

(im guessing im missing a piece of the puzzle that prevents this from happening)

File: teps/0089-nonfalsifiable-provenance-support.md [hunk 4]
```diff
@@ -59,26 +63,41 @@ There are a couple issues with this:
...
+
+Verifying pod identity
+- A pod can request for an trust bundle of a SPIRE server, which will include the public key to validate the pod's certificate.
+- This trust bundle can be used to verify the SVID of a pod
+
+Note: Alternatively, using the JWT SPIFFE endpoint, one can request a JWT with a specified audience field. Although possible, it may not be idiomatic to use the audience for payload signing purposes. However, there is an open [issue](https://github.com/spiffe/spire/issues/1848) discussing adding arbitrary claims as well.
```
Comment (bobcatfish) [inline teps/0089-nonfalsifiable-provenance-support.md:88]: im not familiar with what "audience field" means (or JWT at all TBH X'D) - would it make sense to link to a doc about that here and/or explain why the audience field is relevant in this context?

File: teps/0089-nonfalsifiable-provenance-support.md [hunk 5]
```diff
@@ -59,26 +63,41 @@ There are a couple issues with this:
...
+
+Identity registration is an important aspect of getting SPIRE identities. This defines the subject name of identities that are minted to pods, as well as the attestation requirements of the workloads. This can be done specific to usecase, or there is a kubernetes workload registrar that creates identity based on the fully qualified canonical pod name.
+
+### Using SPIRE for provenance
+
+SPIRE can be used for provenance using the private key provided from the workload's SVID and signing a payload. The signature will be verifiable by the x509 certificate of the SVID together with the trust bundle of the SPIRE deployment. The verification would ensure that the payload's provenance comes from the Tekton entrypointer image (running in Pods) or the Tekton Controller.
```
Comment (bobcatfish) [inline teps/0089-nonfalsifiable-provenance-support.md:94]: nit -> "can be used for signing provenance" ? (i assume we're not saying SPIRE would be creating the provenance)

File: teps/0089-nonfalsifiable-provenance-support.md [hunk 6]
```diff
@@ -59,26 +63,41 @@ There are a couple issues with this:
...
 
 ### Installing SPIRE
-SPIRE runs as a Unix domain socket on the k8s node.
-We can use [spiffe-csi](https://github.com/spiffe/spiffe-csi) to mount the SPIRE socket into Pods as a `csi` type Volume so that we don't have to rely on the `hostPath` volume.
+- SPIRE server runs as a deployment in kubernetes (for simplicity, we'll assume a single cluster SPIRE deployment).
+- SPIRE agents run as a daemonset in the kubernetes cluster, listening on a Unix domain socket on each k8s node.
+- SPIRE kubernetes workload registrar would be optionally installed to provide automatic registration of SPIRE identities. This is optional and could be handled by the Tekton controller as well.
```
Comment (bobcatfish) [inline teps/0089-nonfalsifiable-provenance-support.md:99]: could you explain a bit more about what the registration of SPIRE identities means? (or link to a doc)

File: teps/0089-nonfalsifiable-provenance-support.md [hunk 7]
```diff
@@ -114,8 +133,8 @@ Roughly, this will look something like this:
 
 Each time the TaskRun is modified,
 * Controller verifies the TaskRun hasn't been modified
-* Controller requests a new SPIRE SVID and signature over the new modified TaskRun
-* Controller stores the new SVID and signature on the TaskRun as an annotation
+* Controller requests a new SPIRE SVID signs the new modified TaskRun
```
Comment (bobcatfish) [inline teps/0089-nonfalsifiable-provenance-support.md:136]: typo i think? maybe something like "and then signs" or "and then uses it to sign"

**Review by bobcatfish — COMMENTED** [review]
Thanks for adding more info @lumjjb !! This new info has raised new questions from me, most of which are just nits around wondering what something means - except the new info about how the private key is made available to the caller, im wondering about the implications of that (if any).

File: teps/0089-nonfalsifiable-provenance-support.md (hunk 4 above)
Comment (lumjjb) [reply teps/0089-nonfalsifiable-provenance-support.md:88]: Haha yea for sure! I've spent way more time than I'd like reading this [RFC7519](https://datatracker.ietf.org/doc/html/rfc7519). This [post](https://stackoverflow.com/a/41237822) also gives a good run-down of the practicality around it without standards language.

It kind of is a flexible mechanism that is intended to be used a certain way, which could technically use to sign things but that's not the standard's intention of use.

File: teps/0089-nonfalsifiable-provenance-support.md (hunk 5 above)
Comment (lumjjb) [reply teps/0089-nonfalsifiable-provenance-support.md]: Yup changed!

File: teps/0089-nonfalsifiable-provenance-support.md (hunk 7 above)
Comment (lumjjb) [reply teps/0089-nonfalsifiable-provenance-support.md]: Yup - fixed!

File: teps/0089-nonfalsifiable-provenance-support.md (hunk 6 above)
Comment (lumjjb) [reply teps/0089-nonfalsifiable-provenance-support.md]: Yup - i'll add this link https://spiffe.io/docs/latest/deploying/registering/.

**Review by lumjjb — COMMENTED** [review]
Added some more comments and fixed the typos!

File: teps/0089-nonfalsifiable-provenance-support.md (hunk 3 above)
Comment (bobcatfish) [reply teps/0089-nonfalsifiable-provenance-support.md:82]: Okay thanks for the extra details @lumjjb ! I need a bit of time to chew on this and understand the implications but no reason to block these changes in the meantime!

**Comment by bobcatfish** [comment]
I'd like to chew on the implications of the fact that SPIRE doesn't actually do the signing and that the private key will be available on the node (thanks for answering my questions @lumjjb https://github.com/tektoncd/community/pull/619#discussion_r799589328 ) - and off the cuff I'm thinking we'll want to be explicit in this TEP about how we'll mitigate the potential threat model once we update this to implementable - BUT in the meantime no reason to block this update!

Thanks for adding this extra detail and helping us all improve our understanding @lumjjb 🙏 

/approve

**Pull request merged**